### ISSUE TITLE
Crashing with `java.lang.IllegalStateException: no event up from DESTROYED`

### ISSUE THREAD

**Issue by L-Andrade** [adrielcafe/voyager#280]
~Ever since we upgraded from Voyager 1.0.0-rc05, we've been getting this crash.~
Edit: I misremembered. I believe we've always had this crash since we introduced the Voyager tabs

We're currently on version 1.0.0-rc08, but the most recent updates don't seem to have changed anything that would fix this issue.

The stack trace is:
```
Fatal Exception: java.lang.IllegalStateException: no event up from DESTROYED
       at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.kt:250)
       at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.kt:287)
       at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.kt:136)
       at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.kt:119)
       at cafe.adriel.voyager.androidx.AndroidScreenLifecycleOwner.emitOnStopEvents(AndroidScreenLifecycleOwner.java:107)
       at cafe.adriel.voyager.androidx.AndroidScreenLifecycleOwner.access$emitOnStopEvents(AndroidScreenLifecycleOwner.java:44)
       at cafe.adriel.voyager.androidx.AndroidScreenLifecycleOwner$LifecycleDisposableEffect$1$invoke$$inlined$onDispose$1.dispose(AndroidScreenLifecycleOwner.java:490)
       at androidx.compose.runtime.DisposableEffectImpl.onForgotten(Effects.kt:86)
       at androidx.compose.runtime.CompositionImpl$RememberEventDispatcher.dispatchRememberObservers(Composition.kt:1126)
       at androidx.compose.runtime.CompositionImpl.applyChangesInLocked(Composition.kt:828)
       at androidx.compose.runtime.CompositionImpl.applyChanges(Composition.kt:849)
       at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:625)
       at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:537)
       at androidx.compose.ui.platform.AndroidUiFrameClock$withFrameNanos$2$callback$1.doFrame(AndroidUiFrameClock.android.kt:41)
       at androidx.compose.ui.platform.AndroidUiDispatcher.performFrameDispatch(AndroidUiDispatcher.java:109)
       at androidx.compose.ui.platform.AndroidUiDispatcher.access$performFrameDispatch(AndroidUiDispatcher.java:41)
       at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.doFrame(AndroidUiDispatcher.android.kt:69)
       at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1648)
       at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1659)
       at android.view.Choreographer.doCallbacks(Choreographer.java:1129)
       at android.view.Choreographer.doFrame(Choreographer.java:1045)
       at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1622)
       at android.os.Handler.handleCallback(Handler.java:958)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loopOnce(Looper.java:230)
       at android.os.Looper.loop(Looper.java:319)
       at android.app.ActivityThread.main(ActivityThread.java:8893)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103)
```

We have about ~20-30 crashes every 7 days, which is not a lot considering the size of our userbase, but we would like to fix the issue nonetheless.

Unfortunately, I haven't been able to reproduce the issue at all. I'm not entirely sure how it could happen, and I've tried quite a lot of things. I believe the issue can definitely be in our integration, of course.

We can't share the code, but our navigation is more or less like this:

```
BottomSheetNavigator {
  TabNavigator {
    CurrentTab()
  }
}
```

Each tab has a Navigator as well:
```
Navigator(screen) { navigator ->
        LaunchedEffect(navigator) {
            viewModel.reselected.collectLatest { navigator.popUntilRoot() }
        }
        NavigatorTransition(navigator)
    }
```

All ours tabs are classes and have unique keys. It also happened when the classes were objects and regular `Screen`s. We added `TabDisposable` to our `TabNavigator` as well. The crash happened with and without it too.

It seems to happen mostly when returning to the root of a tab from some other screens, such as:
Initial tab -> Product screen -> ... -> Back to initial tab -> Crash

Do you have any clues why or how this crash could happen? I've tried looking at the source code of `AndroidScreenLifecycleOwner`, and all I could think of would be lifecycle events being handled multiple times when not always necessary due to `DisposableEffect`'s `onDispose` being called.

~This didn't happen when we were using `1.0.0-rc05` though. It started happening once we updated to `1.0.0-rc07`, and is still happening on `1.0.0-rc08`.~ - this has happened since we introduced Voyager tabs in the project.

Might be worth noting that we do use other `Activity`s, and that the `Activity` with the tabs navigator is not the app's "first" `Activity`.

**Comment by DevSrSouza** [comment]
If you could make a reproduceable of this would be awesome!

@Syer10 @adrielcafe 

It seems that we should check the current state before emiting the stopEvents, aparently, it is already `DESTROYED `, I really would want to be able to reproduce to validate a proper fix the code with a workaround :S but maybe a trycatch here or validate the current state would fix this.
```kotlin
private fun emitOnStopEvents() {
        stopEvents.forEach {
            lifecycle.handleLifecycleEvent(it)
        }
    }
```

**Comment by L-Andrade** [comment]
@DevSrSouza Unfortunately I haven't managed to reproduce it yet 😞 

Let me know if you have anything in mind that I can try to replicate it! I'll give it some more tries next week as well once I'm back to work

**Comment by L-Andrade** [comment]
I believe this is due to a DataBinding component we have inside a `Screen`. We were setting the `lifecycleOwner` of this DataBinding component to the `LocalLifecycleOwner` (which is an `AndroidScreenLifecycleOwner`). Inside the generated DataBinding code, there are multiple things happening, including observers being attached to the lifecycle.

This issue is similar, albeit from a different context entirely: https://issuetracker.google.com/issues/184664896

I'm not sure how you folks want to deal with this, or if you want to deal with it at all. In that issue, they just checked the state before emitting it: https://android-review.googlesource.com/c/platform/frameworks/support/+/1673541/4/car/app/app/src/main/java/androidx/car/app/Screen.java#321

The way I managed to reproduce the issue was to make the DataBinding component appear, disappear, and then appear again. I needed some code changes to be able to reproduce it exactly, so I'm not entirely sure it is the only issue.

For now, I've removed the LifecycleOwner attribution from the DataBinding component. I believe it's used more when there's observers/collectors in the DataBinding component, but in this one we don't use it, and it is still working as expected, so it shouldn't be a big issue.

The build with this change will ship on the 15th of January. I'll ping here if the issue appears to be fixed. I might try to replicate it later in an empty project. Feel free to try it yourself as well, of course!

**Comment by L-Andrade** [comment]
@DevSrSouza Here's a project that replicates the issue:

- `MainActivity.kt`:
```kotlin
class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            VoyagerDemoTheme {
                SlideNavigator(InitialScreen())
            }
        }
    }
}

@Composable
private fun SlideNavigator(navScreen: Screen) {
    Navigator(navScreen) { navigator ->
        SlideTransition(navigator)
    }
}


class InitialScreen : Screen {
    override val key: ScreenKey = uniqueScreenKey

    @Composable
    override fun Content() {
        val navigator = LocalNavigator.currentOrThrow
        Button(onClick = { navigator.push(Screen1()) }) {
            Text(text = "Push new Screen")
        }
    }
}

class Screen1 : Screen {
    override val key: ScreenKey = uniqueScreenKey

    @Composable
    override fun Content() {
        Box(modifier = Modifier.fillMaxSize()) {
            var show by remember { mutableStateOf(false) }
            LaunchedEffect(key1 = Unit) {
                while (isActive) {
                    delay(250)
                    show = !show
                }
            }
            if (!show) {
                CircularProgressIndicator(modifier = Modifier.align(Alignment.Center))
            } else {
                val lifecycleOwner = LocalLifecycleOwner.current
                AndroidViewBinding(
                    factory = { inflater, parent, attachToParent ->
                        LayoutDataBindingBinding.inflate(inflater, parent, attachToParent)
                            .apply {
                                this.lifecycleOwner = lifecycleOwner
                            }
                    },
                ) {
                    this.text = "Show: $show"
                }
            }
        }
    }
}
```
- `layout_data_binding.xml`
```xml
<?xml version="1.0" encoding="utf-8"?>
<layout xmlns:android="http://schemas.android.com/apk/res/android">

    <data>
        <variable
            name="text"
            type="String" />
    </data>
    <TextView
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        android:text="@{text}" />

</layout>
```

- Dependencies:
```
dependencies {
    implementation("androidx.core:core-ktx:1.9.0")
    implementation("androidx.lifecycle:lifecycle-runtime-ktx:2.6.1")
    implementation("androidx.activity:activity-compose:1.7.2")
    implementation(platform("androidx.compose:compose-bom:2023.10.01"))
    implementation("androidx.compose.ui:ui")
    implementation("androidx.compose.ui:ui-graphics")
    implementation("androidx.compose.ui:ui-tooling-preview")
    implementation("androidx.compose.material3:material3")
    debugImplementation("androidx.compose.ui:ui-tooling")
    debugImplementation("androidx.compose.ui:ui-test-manifest")
    val voyagerVersion = "1.0.0"
    implementation("cafe.adriel.voyager:voyager-navigator:$voyagerVersion")
    implementation("cafe.adriel.voyager:voyager-transitions:$voyagerVersion")

    implementation("androidx.compose.ui:ui-viewbinding")
}
```

Once setup, you just need to navigate to the next screen and once it hides and shows a few times, navigate back. The crash occurs when navigating back.

Also managed to replicate the issue with this instead of the `AndroidViewBinding`:
```kotlin
val lifecycleOwner = LocalLifecycleOwner.current
LaunchedEffect(key1 = Unit) {
    lifecycleOwner.lifecycle.addObserver(object : DefaultLifecycleObserver {})
}
```

When using a `DisposableEffect` and properly removing the observer in `onDispose`, it did not crash. So it might mean that the DataBinding component is not properly clearing the observer?

Nevertheless, it might be nice if Voyager would be defensive here and only emit when it's supposed to, instead of creating these crashes. They're pretty hard to debug from an app perspective (even if there might be a leak on the app's end).

**Comment by DevSrSouza** [comment]
@L-Andrade release the fix in `1.1.0-alpha02`
